File: screens/ActivationScreen.js
import { useState } from 'react';
import {
 View, Text, TextInput, TouchableOpacity,
 StyleSheet, ActivityIndicator, Alert, Linking
} from 'react-native';
import { signOut } from 'firebase/auth';
import {
 doc, getDoc, updateDoc, addDoc,
 collection, serverTimestamp, setDoc
} from 'firebase/firestore';
import { auth, db } from '../firebase';
import { useTheme } from '../ThemeContext';

const COMMISSION_PCT = 10;
const PLANS = [
 { id:'7d', label:'7 أيام', price:5 },
 { id:'1m', label:'شهر', price:10 },
 { id:'3m', label:'3 أشهر', price:30 },
 { id:'6m', label:'6 أشهر', price:60 },
 { id:'12m', label:'12 شهر', price:100 },
];

export default function ActivationScreen({ user, subStatus }) {
 const { colors, isDark, toggleTheme } = useTheme();
 const [code, setCode] = useState('');
 const [refCode, setRefCode] = useState('');
 const [load, setLoad] = useState(false);

 const activate = async () => {
   const trimmed = code.trim().toUpperCase();
   if (!trimmed) { Alert.alert('خطأ', 'أدخل كود التفعيل'); return; }
   setLoad(true);
   try {
     const ref = doc(db, 'subscriptions', trimmed);
     const snap = await getDoc(ref);
     if (!snap.exists()) { Alert.alert('خطأ', 'الكود غير صحيح'); setLoad(false); return; }
     const data = snap.data();
     if (data.usedBy && data.usedBy !== user.uid) { Alert.alert('خطأ', 'هذا الكود مستخدم مسبقاً'); setLoad(false); return; }
     const exp = data.expiresAt?.toDate ? data.expiresAt.toDate() : new Date(data.expiresAt);
     if (exp < new Date()) { Alert.alert('خطأ', 'هذا الكود منتهي الصلاحية'); setLoad(false); return; }

     // نظام الإحالة
     const refTrimmed = refCode.trim().toUpperCase();
     let bonusMonths = 0;
     if (refTrimmed) {
       const affRef = doc(db, 'affiliates', refTrimmed);
       const affSnap = await getDoc(affRef);
       if (!affSnap.exists()) { Alert.alert('خطأ', 'كود الإحالة غير صحيح'); setLoad(false); return; }
       const affData = affSnap.data();
       const userRef = doc(db, 'users', user.uid);
       const userSnap = await getDoc(userRef);
       if (userSnap.exists() && userSnap.data().usedReferral) {
         Alert.alert('خطأ', 'لقد استخدمت كود إحالة مسبقاً'); setLoad(false); return;
       }
       bonusMonths = 1;
       const plan = PLANS.find(p => p.id === data.plan);
       const commAmount = plan ? Math.round(plan.price * (affData.commissionPct || COMMISSION_PCT) / 100 * 100) / 100 : null;
       await updateDoc(affRef, { totalReferrals: (affData.totalReferrals || 0) + 1 });
       if (commAmount) {
         await addDoc(collection(db, `affiliates/${affSnap.id}/payments`), {
           amount: commAmount,
           note: `عمولة تلقائية — ${user.email}`,
           date: new Date().toISOString().split('T')[0],
           paidBy: 'تلقائي', isAuto: true,
           subscriptionCode: trimmed, createdAt: serverTimestamp()
         });
         await updateDoc(affRef, { totalPaid: (affData.totalPaid || 0) + commAmount });
       }
       await setDoc(doc(db, 'users', user.uid), {
         uid: user.uid, email: user.email,
         usedReferral: true, referralCode: refTrimmed,
         referralUsedAt: serverTimestamp()
       }, { merge: true });
     }

     // تفعيل الكود
     if (bonusMonths > 0) {
       const newExp = new Date(exp);
       newExp.setMonth(newExp.getMonth() + bonusMonths);
       await updateDoc(ref, {
         usedBy: user.uid, usedAt: serverTimestamp(),
         usedByEmail: user.email, expiresAt: newExp,
         planLabel: (data.planLabel || '') + ' + شهر مجاني 🎁'
       });
       Alert.alert('🎉 تم التفعيل', 'مرحباً بك!\nحصلت على شهر مجاني إضافي بفضل كود الإحالة 🎁');
     } else {
       await updateDoc(ref, {
         usedBy: user.uid, usedAt: serverTimestamp(), usedByEmail: user.email
       });
       Alert.alert('✅ تم التفعيل', 'مرحباً بك في إدارة بطاقاتك!');
     }
   } catch (e) { Alert.alert('خطأ', e.message); }
   setLoad(false);
 };

 return (
   <View style={[s.wrap, { backgroundColor: colors.bg }]}>
     <TouchableOpacity style={[s.themeBtn, { backgroundColor: colors.card, borderColor: colors.border }]} onPress={toggleTheme}>
       <Text style={s.themeBtnT}>{isDark ? '☀️' : '🌙'}</Text>
     </TouchableOpacity>

     <Text style={s.logo}>💳</Text>
     <Text style={[s.title, { color: colors.text }]}>تفعيل الاشتراك</Text>
     <Text style={[s.sub, { color: colors.textMuted }]}>{user.email}</Text>

     {subStatus === 'expired' && (
       <View style={s.expBox}>
         <Text style={s.expT}>⚠️ انتهى اشتراكك — تواصل مع المسؤول للتجديد</Text>
       </View>
     )}

     <View style={[s.card, { backgroundColor: colors.card, borderColor: colors.border }]}>
       <Text style={[s.label, { color: colors.textSoft }]}>كود التفعيل *</Text>
       <TextInput style={[s.input, { backgroundColor: colors.inputBg, borderColor: colors.border, color: colors.text }]}
         placeholder="XXXXXXXX"
         placeholderTextColor={colors.textMuted}
         value={code}
         onChangeText={v => setCode(v.toUpperCase())}
         autoCapitalize="characters"
         autoCorrect={false}
         maxLength={12} />

       <Text style={[s.label, { color: colors.textSoft, marginTop:12 }]}>
         كود الإحالة
         <Text style={{color:'#2ecc71', fontWeight:'400'}}> (اختياري — للحصول على شهر مجاني 🎁)</Text>
       </Text>
       <TextInput style={[s.input, { backgroundColor: colors.inputBg, borderColor: colors.border, color: colors.text }]}
         placeholder="مثال: MOHAMAD47"
         placeholderTextColor={colors.textMuted}
         value={refCode}
         onChangeText={v => setRefCode(v.toUpperCase())}
         autoCapitalize="characters"
         autoCorrect={false} />

       {refCode.length > 0 && (
         <View style={s.bonusBox}>
           <Text style={s.bonusT}>🎁 ستحصل على شهر مجاني إضافي!</Text>
         </View>
       )}

       <TouchableOpacity style={[s.btn, { backgroundColor: colors.gold }]} onPress={activate} disabled={load}>
         {load ? <ActivityIndicator color={colors.bg} /> : <Text style={[s.btnT, { color: colors.bg }]}>🔓 تفعيل الاشتراك</Text>}
       </TouchableOpacity>
     </View>

     <TouchableOpacity style={s.waBtn}
       onPress={() => Linking.openURL('https://wa.me/218945888844')}>
       <Text style={s.waBtnT}>📱 تواصل مع خدمة العملاء عبر واتساب</Text>
     </TouchableOpacity>

     <TouchableOpacity onPress={() => signOut(auth)}>
       <Text style={[s.link, { color: colors.textMuted }]}>تسجيل خروج</Text>
     </TouchableOpacity>
   </View>
 );
}

const s = StyleSheet.create({
 wrap:     { flex:1, alignItems:'center', justifyContent:'center', padding:24 },
 themeBtn: { position:'absolute', top:50, left:16, width:40, height:40, borderRadius:20, borderWidth:1, alignItems:'center', justifyContent:'center', zIndex:10 },
 themeBtnT:{ fontSize:18 },
 logo:     { fontSize:50, marginBottom:12 },
 title:    { fontSize:22, fontWeight:'900', marginBottom:4 },
 sub:      { fontSize:13, marginBottom:20 },
 expBox:   { backgroundColor:'rgba(231,76,60,0.1)', borderWidth:1, borderColor:'rgba(231,76,60,0.3)', borderRadius:10, padding:12, marginBottom:16, width:'100%' },
 expT:     { color:'#ff8a80', fontSize:13, textAlign:'center' },
 card:     { borderRadius:16, padding:24, width:'100%', borderWidth:1, marginBottom:16 },
 label:    { fontSize:13, marginBottom:8, fontWeight:'600' },
 input:    { borderWidth:2, borderRadius:12, padding:14, fontSize:18, textAlign:'center', letterSpacing:3, marginBottom:8 },
 bonusBox: { backgroundColor:'rgba(46,204,113,0.1)', borderWidth:1, borderColor:'rgba(46,204,113,0.3)', borderRadius:8, padding:10, marginBottom:8, alignItems:'center' },
 bonusT:   { color:'#2ecc71', fontSize:13, fontWeight:'700' },
 btn:      { borderRadius:10, padding:14, alignItems:'center', marginTop:8 },
 btnT:     { fontSize:15, fontWeight:'700' },
 waBtn:    { backgroundColor:'rgba(37,211,102,0.1)', borderWidth:1, borderColor:'rgba(37,211,102,0.3)', borderRadius:12, padding:14, width:'100%', alignItems:'center', marginBottom:14 },
 waBtnT:   { color:'#25D366', fontSize:14, fontWeight:'700' },
 link:     { fontSize:13, textDecorationLine:'underline' },
});

File: flutter_app/lib/screens/activation_screen.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:url_launcher/url_launcher.dart';

import '../models/subscription.dart';
import '../theme/app_theme.dart';
import '../widgets/common.dart';
import 'auth_screen.dart' show whatsappUrl;

const commissionPctDefault = 10;

class ActivationScreen extends StatefulWidget {
  final User user;
  final String subStatus;
  const ActivationScreen({super.key, required this.user, required this.subStatus});

  @override
  State<ActivationScreen> createState() => _ActivationScreenState();
}

class _ActivationScreenState extends State<ActivationScreen> {
  final codeCtrl = TextEditingController();
  final refCtrl = TextEditingController();
  bool loading = false;

  @override
  void initState() {
    super.initState();
    refCtrl.addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    codeCtrl.dispose();
    refCtrl.dispose();
    super.dispose();
  }

  void _showMsg(String title, String msg) {
    showDialog(
      context: context,
      builder: (_) => AlertDialog(
        title: Text(title),
        content: Text(msg),
        actions: [TextButton(onPressed: () => Navigator.pop(context), child: const Text('حسناً'))],
      ),
    );
  }

  Future<void> _activate() async {
    final trimmed = codeCtrl.text.trim().toUpperCase();
    if (trimmed.isEmpty) {
      _showMsg('خطأ', 'أدخل كود التفعيل');
      return;
    }
    setState(() => loading = true);
    final db = FirebaseFirestore.instance;
    final user = widget.user;
    try {
      final ref = db.collection('subscriptions').doc(trimmed);
      final snap = await ref.get();
      if (!snap.exists) {
        _showMsg('خطأ', 'الكود غير صحيح');
        setState(() => loading = false);
        return;
      }
      final data = snap.data()!;
      final usedBy = data['usedBy'] as String?;
      if (usedBy != null && usedBy != user.uid) {
        _showMsg('خطأ', 'هذا الكود مستخدم مسبقاً');
        setState(() => loading = false);
        return;
      }
      final expTs = data['expiresAt'] as Timestamp?;
      final exp = expTs?.toDate() ?? DateTime.now();
      if (exp.isBefore(DateTime.now())) {
        _showMsg('خطأ', 'هذا الكود منتهي الصلاحية');
        setState(() => loading = false);
        return;
      }

      final refTrimmed = refCtrl.text.trim().toUpperCase();
      var bonusMonths = 0;

      if (refTrimmed.isNotEmpty) {
        final affRef = db.collection('affiliates').doc(refTrimmed);
        final affSnap = await affRef.get();
        if (!affSnap.exists) {
          _showMsg('خطأ', 'كود الإحالة غير صحيح');
          setState(() => loading = false);
          return;
        }
        final affData = affSnap.data()!;
        final userRef = db.collection('users').doc(user.uid);
        final userSnap = await userRef.get();
        if (userSnap.exists && (userSnap.data()?['usedReferral'] == true)) {
          _showMsg('خطأ', 'لقد استخدمت كود إحالة مسبقاً');
          setState(() => loading = false);
          return;
        }
        bonusMonths = 1;
        final plan = subPlans.where((p) => p.id == data['plan']).firstOrNull;
        final commPct = (affData['commissionPct'] as num?)?.toDouble() ?? commissionPctDefault;
        final commAmount = plan != null ? ((plan.price * commPct / 100) * 100).round() / 100 : null;
        await affRef.update({'totalReferrals': ((affData['totalReferrals'] as num?) ?? 0) + 1});
        if (commAmount != null) {
          await db.collection('affiliates').doc(affSnap.id).collection('payments').add({
            'amount': commAmount,
            'note': 'عمولة تلقائية — ${user.email}',
            'date': DateTime.now().toIso8601String().split('T').first,
            'paidBy': 'تلقائي',
            'isAuto': true,
            'subscriptionCode': trimmed,
            'createdAt': FieldValue.serverTimestamp(),
          });
          await affRef.update({'totalPaid': ((affData['totalPaid'] as num?) ?? 0) + commAmount});
        }
        await userRef.set({
          'uid': user.uid,
          'email': user.email,
          'usedReferral': true,
          'referralCode': refTrimmed,
          'referralUsedAt': FieldValue.serverTimestamp(),
        }, SetOptions(merge: true));
      }

      if (bonusMonths > 0) {
        final newExp = DateTime(exp.year, exp.month + bonusMonths, exp.day, exp.hour, exp.minute, exp.second);
        await ref.update({
          'usedBy': user.uid,
          'usedAt': FieldValue.serverTimestamp(),
          'usedByEmail': user.email,
          'expiresAt': Timestamp.fromDate(newExp),
          'planLabel': '${data['planLabel'] ?? ''} + شهر مجاني 🎁',
        });
        _showMsg('🎉 تم التفعيل', 'مرحباً بك!\nحصلت على شهر مجاني إضافي بفضل كود الإحالة 🎁');
      } else {
        await ref.update({
          'usedBy': user.uid,
          'usedAt': FieldValue.serverTimestamp(),
          'usedByEmail': user.email,
        });
        _showMsg('✅ تم التفعيل', 'مرحباً بك في إدارة بطاقاتك!');
      }
    } catch (e) {
      _showMsg('خطأ', e.toString());
    }
    if (mounted) setState(() => loading = false);
  }

  @override
  Widget build(BuildContext context) {
    final theme = context.watch<ThemeController>();
    final colors = theme.colors;

    return Scaffold(
      backgroundColor: colors.bg,
      body: SafeArea(
        child: Stack(
          children: [
            Center(
              child: SingleChildScrollView(
                padding: const EdgeInsets.all(24),
                child: Column(
                  children: [
                    const Text('💳', style: TextStyle(fontSize: 50)),
                    const SizedBox(height: 12),
                    Text('تفعيل الاشتراك', style: TextStyle(fontSize: 22, fontWeight: FontWeight.w900, color: colors.text)),
                    const SizedBox(height: 4),
                    Text(widget.user.email ?? '', style: TextStyle(fontSize: 13, color: colors.textMuted)),
                    const SizedBox(height: 20),
                    if (widget.subStatus == 'expired') const MessageBanner(text: '⚠️ انتهى اشتراكك — تواصل مع المسؤول للتجديد'),
                    Container(
                      width: double.infinity,
                      padding: const EdgeInsets.all(24),
                      decoration: BoxDecoration(
                        color: colors.card,
                        borderRadius: BorderRadius.circular(16),
                        border: Border.all(color: colors.border),
                      ),
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.stretch,
                        children: [
                          ThemedField(
                            label: 'كود التفعيل *',
                            controller: codeCtrl,
                            colors: colors,
                            hint: 'XXXXXXXX',
                            capitalization: TextCapitalization.characters,
                          ),
                          ThemedField(
                            label: 'كود الإحالة (اختياري — للحصول على شهر مجاني 🎁)',
                            controller: refCtrl,
                            colors: colors,
                            hint: 'مثال: MOHAMAD47',
                            capitalization: TextCapitalization.characters,
                          ),
                          if (refCtrl.text.isNotEmpty)
                            const MessageBanner(text: '🎁 ستحصل على شهر مجاني إضافي!', isError: false),
                          const SizedBox(height: 4),
                          PrimaryButton(label: '🔓 تفعيل الاشتراك', onPressed: _activate, loading: loading, colors: colors),
                        ],
                      ),
                    ),
                    const SizedBox(height: 16),
                    SizedBox(
                      width: double.infinity,
                      child: OutlinedButton(
                        style: OutlinedButton.styleFrom(
                          backgroundColor: const Color(0x1A25D366),
                          side: const BorderSide(color: Color(0x4D25D366)),
                          padding: const EdgeInsets.symmetric(vertical: 14),
                          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
                        ),
                        onPressed: () => launchUrl(Uri.parse(whatsappUrl), mode: LaunchMode.externalApplication),
                        child: const Text('📱 تواصل مع خدمة العملاء عبر واتساب',
                            style: TextStyle(color: Color(0xFF25D366), fontSize: 14, fontWeight: FontWeight.w700)),
                      ),
                    ),
                    const SizedBox(height: 14),
                    TextButton(
                      onPressed: () => FirebaseAuth.instance.signOut(),
                      child: Text('تسجيل خروج',
                          style: TextStyle(color: colors.textMuted, fontSize: 13, decoration: TextDecoration.underline)),
                    ),
                  ],
                ),
              ),
            ),
            Positioned(
              top: 8,
              left: 8,
              child: IconButton(
                onPressed: theme.toggle,
                style: IconButton.styleFrom(backgroundColor: colors.card, side: BorderSide(color: colors.border)),
                icon: Text(theme.isDark ? '☀️' : '🌙'),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull => isEmpty ? null : first;
}
